server/store/datastore/migration/000_legacy_to_xorm.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2021 Woodpecker Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package migration
+
+import (
+	"fmt"
+
+	"github.com/rs/zerolog/log"
+	"xorm.io/xorm"
+	"xorm.io/xorm/schemas"
+)
+
+var legacy2Xorm = task{
+	name: "xorm",
+	fn: func(sess *xorm.Session) error {
+		// make sure we have required migrations - else fail and point to last major version
+		for _, mig := range []string{
+			// users
+			"create-table-users",
+			"update-table-set-users-token-and-secret-length",
+			// repos
+			"create-table-repos",
+			"alter-table-add-repo-visibility",
+			"update-table-set-repo-visibility",
+			"alter-table-add-repo-seq",
+			"update-table-set-repo-seq",
+			"update-table-set-repo-seq-default",
+			"alter-table-add-repo-active",
+			"update-table-set-repo-active",
+			"alter-table-add-repo-fallback", // needed to drop col
+			// builds
+			"create-table-builds",
+			"create-index-builds-repo",
+			"create-index-builds-author",
+			// procs
+			"create-table-procs",
+			"create-index-procs-build",
+			// files
+			"create-table-files",
+			"create-index-files-builds",
+			"create-index-files-procs",
+			"alter-table-add-file-pid",
+			"alter-table-add-file-meta-passed",
+			"alter-table-add-file-meta-failed",
+			"alter-table-add-file-meta-skipped",
+			"alter-table-update-file-meta",
+			// secrets
+			"create-table-secrets",
+			"create-index-secrets-repo",
+			// registry
+			"create-table-registry",
+			"create-index-registry-repo",
+			// senders
+			"create-table-senders",
+			"create-index-sender-repos",
+			// perms
+			"create-table-perms",
+			"create-index-perms-repo",
+			"create-index-perms-user",
+			// build_config
+			"create-table-build-config",
+			"populate-build-config",
+		} {
+			exist, err := sess.Exist(&migrations{mig})
+			if err != nil {
+				return fmt.Errorf("test migration existence: %v", err)
+			}
+			if !exist {
+				log.Error().Msgf("migration step '%s' missing, please upgrade to last stable v0.14.x version first", mig)
```

Comment:
Does this mean new user can not start with 0.15.0 directly, but have to use 0.14.x first to setup a clean database?

Reply:
No, clean setup works bit different now:

It just use Sync2 on all needed structs & dummy add migrations.